[Dynamo][Guards]Complete user stack recompilation reason post-processing

fxdawnn opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #170841

Fix https://github.com/pytorch/pytorch/issues/167898 Continuing of PR #169999 to add user stack trace to recompilation reason.

Problem and solution motivation see above PR. This PR focus on adding the refined user stack trace and post process it and only adding it to the recompilation reason. This will provide better debuggability as users are able to see the stack trace for the guard being recompiled. Meanwhile, our design focus on not compromising guards performance.

- The example results in recompile log:
<img width="999" height="241" alt="Screenshot 2025-12-30 at 12 13 14 PM" src="https://github.com/user-attachments/assets/262348e0-cae7-4393-9049-d1ebad028eac" />



- The TL parse example with the user stack only exists in the recompile reason: https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/.tmpdjqYec/dedicated_log_torch_trace_5beo_v54.log/-_0_1_0/recompile_reasons_3.json?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000 (require Meta internal access)
<img width="1131" height="168" alt="Screenshot 2025-12-30 at 12 14 09 PM" src="https://github.com/user-attachments/assets/da37021b-a8e9-40ed-8d52-7758e92af2c4" />




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo

fxdawnn committed:
[Dynamo][Guards]Complete user stack recompilation reason post-processing

[ghstack-poisoned]

fxdawnn committed:
Update on "[Dynamo][Guards]Complete user stack recompilation reason post-processing"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

williamwen42 reviewed (commented):
Will you also update tlparse to render the recompile reason json in a more pleasant format?

fxdawnn commented on torch/_dynamo/guards.py:
```diff
@@ -4479,6 +4498,12 @@ def get_guard_fail_reason_helper(
+
+        # Format user stack trace if available and recompile logging is enabled
+        if (
+            is_recompiles_enabled() or is_recompiles_verbose_enabled()
```
the verbose is where the tlparse initiated as well.

Lucaskabela commented on torch/_dynamo/guards.py:
```diff
@@ -4524,7 +4549,7 @@ def get_guard_fail_reason_helper(
                     break
 
-    reason_str = f"{compile_id}: " + "; ".join(reasons)
+    reason_str = f"{compile_id}: " + "; ".join(reasons) + user_stack_str
```
Nvm answered my own question looking at the output

fxdawnn committed:
Update on "[Dynamo][Guards]Complete user stack recompilation reason post-processing"


Fix https://github.com/pytorch/pytorch/issues/167898 Continuing of PR #169999 to add user stack trace to recompilation reason.

Problem and solution motivation see above PR. This PR focus on adding the refined user stack trace and post process it and only adding it to the recompilation reason. This will provide better debuggability as users are able to see the stack trace for the guard being recompiled. Meanwhile, our design focus on not compromising guards performance.

- The example results in recompile log:
<img width="481" height="221" alt="Screenshot 2025-12-18 at 10 53 45 PM" src="https://github.com/user-attachments/assets/965b4870-0090-4dd3-bfb0-4de7ca5a0aa5" />

- The TL parse example with the user stack only exists in the recompile reason: https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/.tmpAH05y5/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000 (require Meta internal access)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

fxdawnn committed:
Update on "[Dynamo][Guards]Complete user stack recompilation reason post-processing"


Fix https://github.com/pytorch/pytorch/issues/167898 Continuing of PR #169999 to add user stack trace to recompilation reason.

Problem and solution motivation see above PR. This PR focus on adding the refined user stack trace and post process it and only adding it to the recompilation reason. This will provide better debuggability as users are able to see the stack trace for the guard being recompiled. Meanwhile, our design focus on not compromising guards performance.

- The example results in recompile log:
<img width="544" height="207" alt="Screenshot 2025-12-22 at 1 29 31 PM" src="https://github.com/user-attachments/assets/beca2ff3-c7b0-4c6b-b326-e1f1397e79dd" />


- The TL parse example with the user stack only exists in the recompile reason: https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/.tmpOiHYJl/dedicated_log_torch_trace_ushwt809.log/-_0_1_0/recompile_reasons_3.json?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000 (require Meta internal access)
<img width="1501" height="95" alt="Screenshot 2025-12-22 at 1 26 37 PM" src="https://github.com/user-attachments/assets/89372a1b-c4fb-4583-a259-988470677217" />



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

fxdawnn commented on test/dynamo/test_logging.py:
```diff
@@ -197,13 +197,51 @@ def test_cudagraphs(self, records):
+        self.assertIn(
+            """1: {"name": "outer_fn", "line": "return fn(x, ys, zs)"}""",
+            record_str,
+        )
```
yes. changed to check all printed output.

fxdawnn committed:
Update on "[Dynamo][Guards]Complete user stack recompilation reason post-processing"


Fix https://github.com/pytorch/pytorch/issues/167898 Continuing of PR #169999 to add user stack trace to recompilation reason.

Problem and solution motivation see above PR. This PR focus on adding the refined user stack trace and post process it and only adding it to the recompilation reason. This will provide better debuggability as users are able to see the stack trace for the guard being recompiled. Meanwhile, our design focus on not compromising guards performance.

- The example results in recompile log:
<img width="676" height="174" alt="Screenshot 2025-12-29 at 1 55 02 PM" src="https://github.com/user-attachments/assets/e7f0294b-5b84-4051-8055-edcd663bafee" />



- The TL parse example with the user stack only exists in the recompile reason: https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/.tmpjU0CFd/dedicated_log_torch_trace_bixlbv9n.log/-_0_1_0/recompile_reasons_3.json?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000 (require Meta internal access)
<img width="1222" height="122" alt="Screenshot 2025-12-29 at 1 53 26 PM" src="https://github.com/user-attachments/assets/e75efea1-514d-4af0-99f1-66db950af264" />



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

fxdawnn commented:
The content in recompilation reason is primarily fixed with only sending the singly existing recompilation reason through ``payload_fn``. The details display can be accessed at the top of the PR description.

fxdawnn committed:
Update on "[Dynamo][Guards]Complete user stack recompilation reason post-processing"


Fix https://github.com/pytorch/pytorch/issues/167898 Continuing of PR #169999 to add user stack trace to recompilation reason.

Problem and solution motivation see above PR. This PR focus on adding the refined user stack trace and post process it and only adding it to the recompilation reason. This will provide better debuggability as users are able to see the stack trace for the guard being recompiled. Meanwhile, our design focus on not compromising guards performance.

- The example results in recompile log:
<img width="676" height="174" alt="Screenshot 2025-12-29 at 1 55 02 PM" src="https://github.com/user-attachments/assets/e7f0294b-5b84-4051-8055-edcd663bafee" />



- The TL parse example with the user stack only exists in the recompile reason: https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/.tmpjU0CFd/dedicated_log_torch_trace_bixlbv9n.log/-_0_1_0/recompile_reasons_3.json?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000 (require Meta internal access)
<img width="1222" height="122" alt="Screenshot 2025-12-29 at 1 53 26 PM" src="https://github.com/user-attachments/assets/e75efea1-514d-4af0-99f1-66db950af264" />



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

fxdawnn committed:
Update on "[Dynamo][Guards]Complete user stack recompilation reason post-processing"


Fix https://github.com/pytorch/pytorch/issues/167898 Continuing of PR #169999 to add user stack trace to recompilation reason.

Problem and solution motivation see above PR. This PR focus on adding the refined user stack trace and post process it and only adding it to the recompilation reason. This will provide better debuggability as users are able to see the stack trace for the guard being recompiled. Meanwhile, our design focus on not compromising guards performance.

- The example results in recompile log:
<img width="676" height="174" alt="Screenshot 2025-12-29 at 1 55 02 PM" src="https://github.com/user-attachments/assets/e7f0294b-5b84-4051-8055-edcd663bafee" />



- The TL parse example with the user stack only exists in the recompile reason: https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/.tmpjU0CFd/dedicated_log_torch_trace_bixlbv9n.log/-_0_1_0/recompile_reasons_3.json?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000 (require Meta internal access)
<img width="1222" height="122" alt="Screenshot 2025-12-29 at 1 53 26 PM" src="https://github.com/user-attachments/assets/e75efea1-514d-4af0-99f1-66db950af264" />



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

fxdawnn committed:
Update on "[Dynamo][Guards]Complete user stack recompilation reason post-processing"


Fix https://github.com/pytorch/pytorch/issues/167898 Continuing of PR #169999 to add user stack trace to recompilation reason.

Problem and solution motivation see above PR. This PR focus on adding the refined user stack trace and post process it and only adding it to the recompilation reason. This will provide better debuggability as users are able to see the stack trace for the guard being recompiled. Meanwhile, our design focus on not compromising guards performance.

- The example results in recompile log:
<img width="676" height="174" alt="Screenshot 2025-12-29 at 1 55 02 PM" src="https://github.com/user-attachments/assets/e7f0294b-5b84-4051-8055-edcd663bafee" />



- The TL parse example with the user stack only exists in the recompile reason: https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/.tmpjU0CFd/dedicated_log_torch_trace_bixlbv9n.log/-_0_1_0/recompile_reasons_3.json?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000 (require Meta internal access)
<img width="1222" height="122" alt="Screenshot 2025-12-29 at 1 53 26 PM" src="https://github.com/user-attachments/assets/e75efea1-514d-4af0-99f1-66db950af264" />



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

ezyang commented:
I am just looking at the screenshot. It doesn't look like a stack trace to me.

fxdawnn committed:
Update on "[Dynamo][Guards]Complete user stack recompilation reason post-processing"


Fix https://github.com/pytorch/pytorch/issues/167898 Continuing of PR #169999 to add user stack trace to recompilation reason.

Problem and solution motivation see above PR. This PR focus on adding the refined user stack trace and post process it and only adding it to the recompilation reason. This will provide better debuggability as users are able to see the stack trace for the guard being recompiled. Meanwhile, our design focus on not compromising guards performance.

- The example results in recompile log:
<img width="999" height="241" alt="Screenshot 2025-12-30 at 12 13 14 PM" src="https://github.com/user-attachments/assets/262348e0-cae7-4393-9049-d1ebad028eac" />



- The TL parse example with the user stack only exists in the recompile reason: https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/.tmpdjqYec/dedicated_log_torch_trace_5beo_v54.log/-_0_1_0/recompile_reasons_3.json?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000 (require Meta internal access)
<img width="1131" height="168" alt="Screenshot 2025-12-30 at 12 14 09 PM" src="https://github.com/user-attachments/assets/da37021b-a8e9-40ed-8d52-7758e92af2c4" />




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

fxdawnn commented:
> I am just looking at the screenshot. It doesn't look like a stack trace to me.

That's a good point. I've updated it to be stack trace style.

atalman commented:
@pytorchmergebot revert -c nosignal -m "test/dynamo/test_error_messages.py::ErrorMessagesTest::test_recompile_limit_hit_message [GH job link](https://github.com/pytorch/pytorch/actions/runs/20670661865/job/59351305539) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/fe33066516daa39881a99b236f2fe98b38c4c488)"

fxdawnn commented:
@pytorchbot merge -r

fxdawnn commented:
@pytorchbot rebase

fxdawnn committed:
Update on "[Dynamo][Guards]Complete user stack recompilation reason post-processing"


Fix https://github.com/pytorch/pytorch/issues/167898 Continuing of PR #169999 to add user stack trace to recompilation reason.

Problem and solution motivation see above PR. This PR focus on adding the refined user stack trace and post process it and only adding it to the recompilation reason. This will provide better debuggability as users are able to see the stack trace for the guard being recompiled. Meanwhile, our design focus on not compromising guards performance.

- The example results in recompile log:
<img width="999" height="241" alt="Screenshot 2025-12-30 at 12 13 14 PM" src="https://github.com/user-attachments/assets/262348e0-cae7-4393-9049-d1ebad028eac" />



- The TL parse example with the user stack only exists in the recompile reason: https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/.tmpdjqYec/dedicated_log_torch_trace_5beo_v54.log/-_0_1_0/recompile_reasons_3.json?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000 (require Meta internal access)
<img width="1131" height="168" alt="Screenshot 2025-12-30 at 12 14 09 PM" src="https://github.com/user-attachments/assets/da37021b-a8e9-40ed-8d52-7758e92af2c4" />




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

fxdawnn committed:
Update on "[Dynamo][Guards]Complete user stack recompilation reason post-processing"


Fix https://github.com/pytorch/pytorch/issues/167898 Continuing of PR #169999 to add user stack trace to recompilation reason.

Problem and solution motivation see above PR. This PR focus on adding the refined user stack trace and post process it and only adding it to the recompilation reason. This will provide better debuggability as users are able to see the stack trace for the guard being recompiled. Meanwhile, our design focus on not compromising guards performance.

- The example results in recompile log:
<img width="999" height="241" alt="Screenshot 2025-12-30 at 12 13 14 PM" src="https://github.com/user-attachments/assets/262348e0-cae7-4393-9049-d1ebad028eac" />



- The TL parse example with the user stack only exists in the recompile reason: https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/.tmpdjqYec/dedicated_log_torch_trace_5beo_v54.log/-_0_1_0/recompile_reasons_3.json?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000 (require Meta internal access)
<img width="1131" height="168" alt="Screenshot 2025-12-30 at 12 14 09 PM" src="https://github.com/user-attachments/assets/da37021b-a8e9-40ed-8d52-7758e92af2c4" />




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]
